File: prototype/rails/src/test/models/customer_merge_plan_test.rb
require "test_helper"

class CustomerMergePlanTest < ActiveSupport::TestCase
  test "quantities of the same listing on both sides are summed" do
    lines = CustomerMergePlan.fold_cart_lines({ 1 => 2 }, { 1 => 3 }, {})

    assert_equal [ CustomerMergePlan::CartLine.new(listing_id: 1, quantity: 5) ], lines
  end

  test "a summed quantity over stock is clamped to what is left" do
    lines = CustomerMergePlan.fold_cart_lines({ 1 => 2 }, { 1 => 3 }, { 1 => 4 })

    assert_equal [ CustomerMergePlan::CartLine.new(listing_id: 1, quantity: 4) ], lines
  end

  test "a line clamped to zero stock is dropped" do
    lines = CustomerMergePlan.fold_cart_lines({ 1 => 2 }, { 1 => 3 }, { 1 => 0 })

    assert_empty lines
  end

  test "a line the stock table has no cap for is kept in full" do
    lines = CustomerMergePlan.fold_cart_lines({}, { 1 => 3 }, {})

    assert_equal [ CustomerMergePlan::CartLine.new(listing_id: 1, quantity: 3) ], lines
  end

  test "lines on only one side are both kept, unclamped by each other" do
    lines = CustomerMergePlan.fold_cart_lines({ 1 => 1 }, { 2 => 1 }, { 1 => 10, 2 => 10 })

    assert_equal [ 1, 2 ], lines.map(&:listing_id).sort
  end

  test "nothing on either side folds to nothing" do
    assert_empty CustomerMergePlan.fold_cart_lines({}, {}, {})
  end

  test "an anonymous favorite the verified customer does not hold moves" do
    plan = CustomerMergePlan.partition_favorites([ 1 ], [ 2 ])

    assert_equal [ 2 ], plan[:move]
    assert_empty plan[:drop]
  end

  test "an anonymous favorite duplicating the verified customer's own drops instead of moving" do
    plan = CustomerMergePlan.partition_favorites([ 1 ], [ 1 ])

    assert_empty plan[:move]
    assert_equal [ 1 ], plan[:drop]
  end

  test "the anonymous side's own duplicate favorites collapse to one decision" do
    plan = CustomerMergePlan.partition_favorites([], [ 1, 1 ])

    assert_equal [ 1 ], plan[:move]
  end

  test "favorites the verified customer already holds and ones only the anonymous customer holds partition together" do
    plan = CustomerMergePlan.partition_favorites([ 1 ], [ 1, 2 ])

    assert_equal [ 2 ], plan[:move]
    assert_equal [ 1 ], plan[:drop]
  end
end

File: prototype/php/src/app/Domain/Customers/CustomerMergePlanTest.php
<?php

declare(strict_types=1);

namespace App\Domain\Customers;

it('folds cart lines', function (
    array $verifiedCartLines,
    array $anonymousCartLines,
    array $stockByListing,
    array $expected,
): void {
    /**
     * @var list<CustomerCartLine> $verifiedCartLines
     * @var list<CustomerCartLine> $anonymousCartLines
     * @var array<string, int> $stockByListing
     */
    $plan = CustomerMergePlan::for($verifiedCartLines, $anonymousCartLines, [], [], $stockByListing);

    expect(array_map(
        fn (CustomerCartLine $line): array => ['listingId' => $line->listingId, 'quantity' => $line->quantity],
        $plan->cartLines,
    ))->toBe($expected);
})->with([
    'disjoint carts merge to the union' => [
        [new CustomerCartLine('lst_1', 2)],
        [new CustomerCartLine('lst_2', 1)],
        [],
        [['listingId' => 'lst_1', 'quantity' => 2], ['listingId' => 'lst_2', 'quantity' => 1]],
    ],
    'the same listing on both sides sums quantities' => [
        [new CustomerCartLine('lst_1', 2)],
        [new CustomerCartLine('lst_1', 3)],
        [],
        [['listingId' => 'lst_1', 'quantity' => 5]],
    ],
    'a summed quantity above stock clamps to stock' => [
        [new CustomerCartLine('lst_1', 2)],
        [new CustomerCartLine('lst_1', 3)],
        ['lst_1' => 4],
        [['listingId' => 'lst_1', 'quantity' => 4]],
    ],
    'a line already at stock is unaffected' => [
        [new CustomerCartLine('lst_1', 4)],
        [],
        ['lst_1' => 4],
        [['listingId' => 'lst_1', 'quantity' => 4]],
    ],
    'stock of zero drops the line' => [
        [new CustomerCartLine('lst_1', 2)],
        [],
        ['lst_1' => 0],
        [],
    ],
    'a listing missing from the stock map is not clamped' => [
        [new CustomerCartLine('lst_1', 50)],
        [],
        [],
        [['listingId' => 'lst_1', 'quantity' => 50]],
    ],
    'both carts empty gives an empty cart' => [
        [], [], [], [],
    ],
    'one side empty carries the other through untouched' => [
        [new CustomerCartLine('lst_1', 2)],
        [],
        [],
        [['listingId' => 'lst_1', 'quantity' => 2]],
    ],
    'line order is verified listings first, then anonymous-only listings' => [
        [new CustomerCartLine('lst_3', 1), new CustomerCartLine('lst_1', 1)],
        [new CustomerCartLine('lst_4', 1), new CustomerCartLine('lst_2', 1)],
        [],
        [
            ['listingId' => 'lst_3', 'quantity' => 1],
            ['listingId' => 'lst_1', 'quantity' => 1],
            ['listingId' => 'lst_4', 'quantity' => 1],
            ['listingId' => 'lst_2', 'quantity' => 1],
        ],
    ],
]);

it('unions favorites', function (
    array $verifiedFavoriteListingIds,
    array $anonymousFavoriteListingIds,
    array $expectedMove,
    array $expectedDrop,
): void {
    /**
     * @var list<string> $verifiedFavoriteListingIds
     * @var list<string> $anonymousFavoriteListingIds
     */
    $plan = CustomerMergePlan::for([], [], $verifiedFavoriteListingIds, $anonymousFavoriteListingIds, []);

    expect($plan->favoritesToMove)->toBe($expectedMove)
        ->and($plan->favoritesToDrop)->toBe($expectedDrop);
})->with([
    'a favorite the verified customer does not already have moves' => [
        ['lst_1', 'lst_2'], ['lst_2', 'lst_3'], ['lst_3'], ['lst_2'],
    ],
    'anonymous favorites de-duplicate before moving' => [
        [], ['lst_1', 'lst_1', 'lst_2'], ['lst_1', 'lst_2'], [],
    ],
    'a verified customer with no favorites moves every anonymous favorite' => [
        [], ['lst_1', 'lst_2'], ['lst_1', 'lst_2'], [],
    ],
    'an anonymous customer with no favorites moves and drops nothing' => [
        ['lst_1', 'lst_2'], [], [], [],
    ],
    'both empty gives nothing to move or drop' => [
        [], [], [], [],
    ],
    'the same listing on both sides drops rather than moves' => [
        ['lst_1'], ['lst_1'], [], ['lst_1'],
    ],
]);

it('folds an empty merge into an empty plan', function (): void {
    $plan = CustomerMergePlan::for([], [], [], [], []);

    expect($plan->cartLines)->toBe([])
        ->and($plan->favoritesToMove)->toBe([])
        ->and($plan->favoritesToDrop)->toBe([]);
});
